src/cryptography/hazmat/backends/openssl/x509.py

```diff
@@ -162,6 +163,37 @@ def public_bytes(self, encoding):
         self._backend.openssl_assert(res == 1)
         return self._backend._read_mem_bio(bio)
 
+    def verify(self, intermediates, trusted_roots, cert_verif_cb):
+
+        g = Graph()
+        all_certs = intermediates + trusted_roots
+        all_certs.append(self)
+        for cert in all_certs:
+            g.add_vertex(cert)
+        for cert in all_certs:
+            issuer_name = cert.issuer
+            issuer = None
+            for issuer_candidate in all_certs:
+                if issuer_candidate.subject == issuer_name:
+                    issuer = issuer_candidate
+                    break
+            if issuer:
+                g.add_edge(cert, issuer)
```

Comment:
I might be missing something, but I believe we do not check the signature of the current cert with the issuer public key here which I believe should guard the `add_edge` call.

Reply:
Also, ouldn't someone submit a leaf cert with a `Fake Top` value in the issuer field and that would immediately verify?

Reply:
I'd think that would do the job instead (roughly tested):

```python
        for cert in all_certs:
            issuer_name = cert.issuer
            for issuer_candidate in all_certs:
                if issuer_candidate.subject == issuer_name:
                    pkey = issuer_candidate.public_key()

                    try:
                        if isinstance(pkey, rsa.RSAPublicKeyWithSerialization):
                            pkey.verify(
                                signature=cert.signature,
                                data=cert.tbs_certificate_bytes,
                                padding=padding.PKCS1v15(),
                                algorithm=cert.signature_hash_algorithm,
                            )
                        elif isinstance(
                                pkey,
                                ec.EllipticCurvePublicKeyWithSerialization):
                            pkey.verify(
                                signature=cert.signature,
                                data=cert.tbs_certificate_bytes,
                                signature_algorithm=cert
                                .signature_hash_algorithm,
                            )
                        elif isinstance(
                                pkey, dsa.DSAPublicKeyWithSerialization):
                            pkey.verify(
                                signature=cert.signature,
                                data=cert.tbs_certificate_bytes,
                                algorithm=cert.signature_hash_algorithm,
                            )
                        else:
                            raise UnsupportedAlgorithm(
                                '{} verification is not implemented'
                                .format(pkey.__class__)
                            )

                        g.add_edge(cert, issuer_candidate)

                    except InvalidSignature:
                        pass
```

Reply:
Definitely so! Thank you for the code sketch!
I decided that this check deserves function on its own and implemented it in #5116. Will update *this* PR when the latter (hopefully) will be merged. (The code requried minor change for EC key, `verify()` in that case requires `signature_algorithm` being instance of `ec.ECDSA`.)